src/main/java/io/github/cottonmc/cotton/gui/widget/WGridPanel.java

```diff
@@ -6,23 +6,82 @@
  * A panel that positions children in a grid.
  */
 public class WGridPanel extends WPanelWithInsets {
+
+	private static final int DEFAULT_GRID_SIZE = 18;
 	/**
 	 * The grid size in pixels.
 	 * Defaults to 18, which is the size of one item slot.
 	 */
-	protected int grid = 18;
+	protected int grid;
+	protected int horizontalGap;
+	protected int verticalGap;
 
 	/**
 	 * Constructs a grid panel with the default grid size.
 	 */
-	public WGridPanel() {}
+	public WGridPanel() {
+		this(DEFAULT_GRID_SIZE);
+	}
 
 	/**
 	 * Constructs a grid panel with a custom grid size.
 	 *
 	 * @param gridSize the grid size in pixels
 	 */
-	public WGridPanel(int gridSize) { this.grid = gridSize; }
+	public WGridPanel(int gridSize) {
+		this(gridSize, 0, 0);
+	}
+
+	/**
+	 * Constructs a grid panel with a custom grid size, and gaps between widgets.
+	 *
+	 * @param gridSize      the grid size in pixels
+	 * @param horizontalGap the horizontal gap between two widgets
+	 * @param verticalGap   the vertical gap between two widgets
+	 */
+	public WGridPanel(int gridSize, int horizontalGap, int verticalGap) {
+		this.grid = gridSize;
+		this.horizontalGap = horizontalGap;
+		this.verticalGap = verticalGap;
+	}
+
+	/**
+	 * Constructs a grid panel with a vertical gap between widgets.
+	 *
+	 * @param verticalGap the vertical gap between two widgets
+	 */
+	public static WGridPanel withVerticalGap(int verticalGap) {
+		return new WGridPanel(DEFAULT_GRID_SIZE, 0, verticalGap);
+	}
+
+	/**
+	 * Constructs a grid panel with a horizontal gap between widgets.
+	 *
+	 * @param horizontalGap the horizontal gap between two widgets
+	 */
+	public static WGridPanel withHorizontalGap(int horizontalGap) {
+		return new WGridPanel(DEFAULT_GRID_SIZE, horizontalGap, 0);
+	}
+
+	/**
+	 * Constructs a grid panel with gaps between widgets.
+	 *
+	 * @param horizontalGap the horizontal gap between two widgets
+	 * @param verticalGap   the vertical gap between two widgets
+	 */
+	public static WGridPanel withGaps(int horizontalGap, int verticalGap) {
+		return new WGridPanel(DEFAULT_GRID_SIZE, horizontalGap, verticalGap);
+	}
```

Comment:
Not needed since there's the `int, int, int` constructor

Reply:
I don't like having users of a class specify some default magic constant, so I'll keep this one, but we can negotiate about the other static constructors. I don't really see a point in removing them, in my mind they don't do any harm, and provide sensible defaults instead.

Reply:
See my other comment (https://github.com/CottonMC/LibGui/pull/170#discussion_r938088710), if the constant is public then the magic constant reasoning doesn't really hold.

I think builder-style setters for the gaps would be more useful instead since you could do `new WGridPanel().setHorizontalGap(4)` for example. Or `new WGridPanel().setGaps(4, 2)`. This would also match the rest of LibGui much better.

Reply:
If i remove the constructor and instead switch to `.set*Gaps` should those setters recompute the positioning of all widgets, or should I rely on this being called before any widgets are added (which ill probably do by emitting a warning, or throwing an exception if there are already `children` present)

Reply:
IMO a warning in both the log and javadocs is the best option as re-layouting needs a lot more work.